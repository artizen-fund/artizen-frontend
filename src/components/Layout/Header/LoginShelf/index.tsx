import { useMemo, useState } from 'react'
import { useApolloClient, ApolloClient } from '@apollo/client'
import type { Magic } from 'magic-sdk'
import Link from 'next/link'
import styled from 'styled-components'
import { Button, Icon, Form, CheckboxControl } from '@components'
import { rgba, createSession, useMagic } from '@lib'
import { palette, typography, breakpoint } from '@theme'
import { schema, uischema, initialState, FormState } from './form'

const LoginShelf = () => {
  const LOCALSTORAGE_KEY = 'loginForm'
  const { magic } = useMagic()
  const apolloClient = useApolloClient()

  const [data, setData] = useState<FormState>(initialState)
  useMemo(() => {
    if (typeof localStorage === 'undefined') {
      return
    }
    const frozenAnswers = localStorage.getItem(LOCALSTORAGE_KEY)
    if (!frozenAnswers) {
      setData(initialState)
      return
    }
    const thawedAnswers = JSON.parse(frozenAnswers)
    setData(thawedAnswers)
  }, [])

  const [submitted, setSubmitted] = useState(false)
  const [readonly, setReadonly] = useState(false)
  const [acceptedToc, setAcceptedToc] = useState(true)


  const handleLogin = async (
    magic: Magic, 
    apolloClient: ApolloClient<object>,
    provider?: string, 
    email?: string,
  ) => {
    setReadonly(true)
    setSubmitted(true)

    const loginWithEmail = email ? { email,magic } : undefined
    const loginWithSocial = provider ? {provider,magic } : undefined

    

    try {
      createSession(
        apolloClient,
        loginWithEmail,
        loginWithSocial,
      )
      setSubmitted(true)
      setReadonly(false)
    } catch (error) {
      console.error(error)
      setReadonly(false)
    }
  }

  const reset = () => {
    setSubmitted(false)
    setReadonly(false)
  }

  return (
    <Wrapper className={submitted ? 'submitted' : ''}>
      <Copy>
        <Headline>Let’s get started by setting up your Artizen account</Headline>
        <InfoRow>
          <Icon glyph="infoLarge" outline level={1} />
          <SignInDirections>
            Already have an account?
            <br />
            You can still use this form, it’s magic!
          </SignInDirections>
        </InfoRow>
      </Copy>
      <Form localStorageKey={LOCALSTORAGE_KEY} {...{ schema, uischema, initialState, data, setData, readonly }}>
        <SubmitButton
            stretch onClick={
            () => {
              if(!data.email) return
              handleLogin(magic, apolloClient, undefined, data.email)

            }

        } disabled={!data.email || !acceptedToc || readonly}>
          Sign In / Sign Up
        </SubmitButton>
        <Confirmation>
          <Icon glyph="tick" outline level={2} color="moss" />
          <div>
            <h1>Done, confirmation sent!</h1>
            <p>We emailed a magic link to {data.email}. Click the link Sign in or sign up.</p>
          </div>
          <Reset onClick={() => reset()}>Didn’t receive an email?</Reset>
        </Confirmation>
      </Form>
      <Alternatives>
        <OrLine />
        <Buttons>
          <Button level={1} outline onClick={() => alert('derp')} stretch>
            Phone
          </Button>
          <Button
            level={1} outline onClick={

              () => {
                
                handleLogin(magic, apolloClient, 'twitter', undefined)
              } 
            
          } stretch>
            Twitter
          </Button>
          <Button
                level={1} outline onClick={
               () => {
                handleLogin(magic, apolloClient, 'discord', undefined)
              } 

          } stretch>
            Discord
          </Button>
        </Buttons>
      </Alternatives>
      <TocWrapper>
        <TocCheck>
          <CheckboxControl
            data={acceptedToc}
            path="not-used"
            handleChange={() => setAcceptedToc(!acceptedToc)}
            label=""
          />
          <TocMessage>
            I agree to Artizen’s
            <Link href="/toc">
              <a> Terms &amp; Conditions</a>
            </Link>
          </TocMessage>
        </TocCheck>
      </TocWrapper>
    </Wrapper>
  )
}

const SubmitButton = styled(props => <Button {...props} />)`
  grid-area: submit;
`

const Alternatives = styled.div`
  grid-area: alternatives;
`

const Confirmation = styled.div`
  display: none;
  grid-area: confirmation;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 30px;
  h1 {
    ${typography.title.l4}
    color: ${rgba(palette.night)};
    @media (prefers-color-scheme: dark) {
      color: ${rgba(palette.moon)};
    }
  }
  p {
    ${typography.label.l1}
    color: ${rgba(palette.barracuda)};
  }
  text-align: center;
`

const Reset = styled.p`
  cursor: pointer;
  border-bottom: 2px solid ${rgba(palette.night)};
  @media (prefers-color-scheme: dark) {
    border-bottom: 2px solid ${rgba(palette.moon)};
  }
`

const Wrapper = styled.div`
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-template-areas:
      'copy email'
      'copy submit'
      'tocCheck alternatives';
    &.submitted {
      grid-template-areas:
        'copy confirmation'
        'copy confirmation'
        'tocCheck confirmation';
    }
    gap: 10px;

    .vertical-layout,
    .vertical-layout-item {
      display: contents;
    }

    *[id='#/properties/email'] {
      grid-area: email;
    }
  }

  &.submitted {
    *[id='#/properties/email'],
    ${SubmitButton}, ${Alternatives} {
      display: none;
    }
    ${Confirmation} {
      display: flex;
    }
  }
`

const Copy = styled.div`
  grid-area: copy;
`

const Headline = styled.h1`
  ${typography.title.l2};
`

const InfoRow = styled.div`
  display: flex;
  flex-direction: row;
  gap: 10px;
  justify-content: flex-start;
  align-items: flex-start;
  margin: 1em 0 2em 0;
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    margin: 1em 0 0 0;
  }
`

const SignInDirections = styled.p`
  ${typography.label.l1};
`

const OrLine = styled.div`
  position: relative;
  margin: 1em 0;

  text-align: center;
  ${typography.label.l1}
  color: ${rgba(palette.barracuda)};
  text-transform: uppercase;

  &:before {
    content: ' ';
    position: absolute;
    z-index: 0;
    left: 0;
    top: 40%;
    width: 100%;
    height: 1px;
    background: ${rgba(palette.stone)};
    @media (prefers-color-scheme: dark) {
      background: ${rgba(palette.barracuda, 0.64)};
    }
  }
  &:after {
    position: relative;
    z-index: 1;
    content: ' or ';
    margin-inline: auto;
    padding: 0 10px;
    background: ${rgba(palette.white)};
    @media (prefers-color-scheme: dark) {
      background: ${rgba(palette.slate)};
    }
  }
`

const Buttons = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  gap: 10px;
`

const TocWrapper = styled.div`
  display: contents;
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    grid-area: tocCheck;
  }
`

const TocCheck = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    justify-content: start;
  }
  a {
    text-decoration: underline;
    text-decoration-thickness: 2px;
  }
`

const TocMessage = styled.p`
  ${typography.label.l1}
  color: ${rgba(palette.barracuda)};
`

export default LoginShelf
