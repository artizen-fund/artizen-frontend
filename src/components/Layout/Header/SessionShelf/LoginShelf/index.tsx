import { useState } from 'react'
import { useApolloClient, ApolloClient } from '@apollo/client'
import styled from 'styled-components'
import { Icon, Form, CheckboxControl, Button } from '@components'
import { loginWithEmail, useMagic, useFormLocalStorage } from '@lib'
import { breakpoint } from '@theme'
import { schema, uischema, initialState, FormState } from '@forms/login'
import {
  Copy,
  Headline,
  SignInDirections,
  InfoRow,
  Confirmation,
  Reset,
  CheckWrapper,
  Check,
  CheckMessage,
  ISessionShelf,
} from '../_common'

const LoginShelf = ({ setCreateMode }: ISessionShelf) => {
  const LOCALSTORAGE_KEY = 'loginForm'
  const [data, setData] = useFormLocalStorage<FormState>(LOCALSTORAGE_KEY, initialState)

  const { magic } = useMagic()
  const apolloClient = useApolloClient()

  const [sentEmail, setSentEmail] = useState(false)
  const [readonly, setReadonly] = useState(false)
  const [recordEmail, setRecordEmail] = useState(true)

  const handleEmailLogin = async (apolloClient: ApolloClient<object>, email?: string, magic?: MagicInstance) => {
    if (!magic) {
      throw 'Error: magic session not initialized.'
    }
    if (!email) {
      throw 'Error: email is missing'
    }
    setReadonly(true)
    try {
      setSentEmail(true)
      await loginWithEmail(apolloClient, magic, email)
      setReadonly(false)
    } catch (error) {
      console.error(error)
      setReadonly(false)
    }
  }

  const reset = () => {
    setReadonly(false)
  }

  return (
    <Wrapper className={readonly ? 'submitted' : ''}>
      <Copy>
        <Headline>Sign in using the email associated with your Artizen account</Headline>
        <InfoRow onClick={() => setCreateMode(true)}>
          <Icon glyph="infoLarge" outline level={1} />
          <SignInDirections>Sign up for an account instead</SignInDirections>
        </InfoRow>
      </Copy>
      <Form localStorageKey={LOCALSTORAGE_KEY} {...{ schema, uischema, initialState, data, setData, readonly }}>
        <>
          <Buttons>
            <Button stretch onClick={() => handleEmailLogin(apolloClient, data.email, magic)}>
              Sign In
            </Button>
            <Button transparent stretch onClick={() => setCreateMode(true)}>
              Sign Up instead
            </Button>
          </Buttons>

          {sentEmail && (
            <Confirmation>
              <Icon glyph="tick" outline level={2} color="moss" />
              <div>
                <h1>Done, confirmation sent!</h1>
                <p>
                  We emailed a magic link to {data.email}.<br />
                  Click the link sign in.
                </p>
              </div>
              <Reset onClick={() => reset()}>Didnâ€™t receive an email?</Reset>
            </Confirmation>
          )}
        </>
      </Form>
      <CheckWrapper>
        <Check>
          <CheckboxControl
            data={recordEmail}
            path="not-used"
            handleChange={() => setRecordEmail(!recordEmail)}
            label=""
          />
          <CheckMessage>Remember my email address.</CheckMessage>
        </Check>
      </CheckWrapper>
    </Wrapper>
  )
}

const Buttons = styled.div`
  grid-area: buttons;
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
`

const Wrapper = styled.div`
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-template-areas:
      'copy email'
      'copy buttons'
      'tocCheck buttons';
    &.submitted {
      grid-template-areas:
        'copy confirmation'
        'copy confirmation'
        'tocCheck confirmation';
    }
    gap: 10px;

    .vertical-layout,
    .vertical-layout-item {
      display: contents;
    }

    *[id='#/properties/email'] {
      grid-area: email;
    }
  }

  &.submitted {
    *[id='#/properties/email'],
    ${Buttons} {
      display: none;
    }
    ${Confirmation} {
      display: flex;
    }
  }
`

export default LoginShelf
