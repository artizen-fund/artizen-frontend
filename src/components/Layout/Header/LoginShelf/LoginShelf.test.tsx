import renderer from 'react-test-renderer'
import LoginShelf from './'

describe('LoginShelf', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<LoginShelf />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
