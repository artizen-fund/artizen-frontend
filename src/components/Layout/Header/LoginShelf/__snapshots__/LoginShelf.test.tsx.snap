// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginShelf renders correctly 1`] = `
<div>
  <div
    class="sc-9c2c0fc5-4 iCiVJK"
  >
    <div
      class="sc-9c2c0fc5-5 fLeqkg"
    >
      <h1
        class="sc-9c2c0fc5-6 hOScBH"
      >
        Let’s get started by setting up your Artizen account
      </h1>
      <div
        class="sc-9c2c0fc5-7 jXnRuP"
      >
        <div
          class="sc-c089a03b-2 VsSZo"
        >
          <div
            class="sc-c089a03b-0 ja-DffS"
            color="night"
          >
            <div
              class="sc-149d3190-0 iAXlWi"
              color="night"
            />
          </div>
        </div>
        <p
          class="sc-9c2c0fc5-8 hFjcqt"
        >
          Already have an account?
          <br />
          You can still use this form, it’s magic!
        </p>
      </div>
    </div>
    <div
      class="vertical-layout"
    >
      <div
        class="vertical-layout-item"
      >
        <div
          class="sc-2f9d5146-0 jjUILZ"
          id="#/properties/email"
        >
          <div
            class="sc-dccedff6-0 dxDPQx"
          >
            <input
              placeholder="e.g. rene@artizen.fund"
              required=""
              type="text"
              value=""
            />
            <label
              class="sc-ffdb2e6c-0 dXVxRc"
            >
              Email Address
               *
            </label>
          </div>
          <div
            class="sc-93ba83c0-0 bAVzHI"
          >
            is a required property
          </div>
        </div>
      </div>
    </div>
    <button
      class="sc-2a588df4-2 ixnoRf sc-9c2c0fc5-0 hVAiAY  disabled"
      disabled=""
    >
      <span>
        Sign In / Sign Up
      </span>
    </button>
    <div
      class="sc-9c2c0fc5-1 ccMnwl"
    >
      <div
        class="sc-9c2c0fc5-9 cXUljj"
      />
      <div
        class="sc-9c2c0fc5-10 hrMKlC"
      >
        <button
          class="sc-2a588df4-2 kYegxc undefined "
        >
          <span>
            Phone
          </span>
        </button>
        <button
          class="sc-2a588df4-2 kYegxc undefined "
        >
          <span>
            Twitter
          </span>
        </button>
        <button
          class="sc-2a588df4-2 kYegxc undefined "
        >
          <span>
            Discord
          </span>
        </button>
      </div>
    </div>
    <div
      class="sc-9c2c0fc5-11 pMnPu"
    >
      <div
        class="sc-9c2c0fc5-12 bGBpas"
      >
        <label
          class="sc-dd35d48e-4 dCKwdz"
        >
          <div
            class="sc-dd35d48e-0 lmHjtw"
          >
            <input
              checked=""
              class="sc-dd35d48e-1 clTMzK"
              type="checkbox"
            />
            <span
              class="sc-dd35d48e-2 dtceeR"
            />
          </div>
          <span
            class="sc-dd35d48e-3 eAaGSv"
          />
        </label>
        <p
          class="sc-9c2c0fc5-13 dmkgqG"
        >
          I agree to Artizen’s
          <a
            href="/toc"
          >
             Terms & Conditions
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;
